new tutorial for blockchain notebooks

Jiri-Kremser opened the pull request:
- [x] fix the bug with nbextensions (conda failed during the installation, now using pip instead)
- [x] use the explicit tags for docker images instead of `:latest`

Still open question: should I change the ownership from `jkremser/foo` to `radanalyticsio/foo` on GitHub repo / Dockerhub / both?

otherwise it should be gtg (I removed the WIP from the PR label)

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+:page-labels: [Python, Jupyter, spark-notebook]
+:page-layout: application
+:page-menu_template: menu_tutorial_application.html
+:page-description: The blockchain notebooks are examples of how to explore graph data using GraphX and GraphFrames on top of OpenShift using Spark. It uses the real Bitcoin blockchain data to create transaction graph for the analysis.
```
s/The blockchain notebooks/These blockchain notebooks/

s/using Spark/using Apache Spark/

s/create transaction graph/create a transaction graph/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+[[introduction]]
+== Introduction
+
+Cryptocurrencies attract various groups of people. Among other it could be
```
s/Among other it/Among them/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+
+Cryptocurrencies attract various groups of people. Among other it could be
+investors, people from retail, tech enthusiasts, crypto-anarchists, etc.
+We are not going to focus on anything else than the raw technology behind the
```
s/anything else than/anything other than/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+We are not going to focus on anything else than the raw technology behind the
+Blockchain, leaving aside all the ideology and hype that comes with the Bitcoin.
+
+The https://spark.apache.org/graphx/[GraphX] based notebook uses the https://github.com/spark-notebook/spark-notebook[spark-notebook] as the notebook technology,
```
s/notebook uses the/notebook uses this/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+Blockchain, leaving aside all the ideology and hype that comes with the Bitcoin.
+
+The https://spark.apache.org/graphx/[GraphX] based notebook uses the https://github.com/spark-notebook/spark-notebook[spark-notebook] as the notebook technology,
+while the second one uses https://graphframes.github.io/[GraphFrames] and Jupyter notebook. Also the second
```
i think i would follow your pattern from the earlier in the sentence, so something like:

```
while the https://graphframes.github.io/[GraphFrames] based notebook uses project Jupyter.
```

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+A Jupyter/Spark notebook is an interactive compute environment that interleaves
+documentation, code examples, and graphical output.
+
+Similarly as our other notebooks the notebooks are the drivers that interactively
```
i would say

```
Similar to our other notebook examples, these notebooks are the....
```

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+documentation, code examples, and graphical output.
+
+Similarly as our other notebooks the notebooks are the drivers that interactively
+sends the commands to Spark cluster for the evaluation. They can either spawn
```
s/sends the commands to/send the command to the/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+
+Similarly as our other notebooks the notebooks are the drivers that interactively
+sends the commands to Spark cluster for the evaluation. They can either spawn
+its own standalone Spark or connect to an external one.
```
s/its/their/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+oc expose svc/bitcoin-notebook
+....
+
+for the Jupyter notebook and
```
i would say it like this:

```
for the GraphFrames notebook and the following for the GraphX notebook:
```

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+oc expose svc/bitcoin-spark-notebook
+....
+
+Once both routes are exposed the both notebooks should be available. You
```
s/exposed the both/exposed, the/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+Once both routes are exposed the both notebooks should be available. You
+can find the URLs if you run `oc get routes`.
+It is also possible for the Jupyter notebook to set
+environmental variable that assures it will connect to an external Spark cluster.
```
s/assures/assure/

Jiri-Kremser replied:
I chose different wording here, but why can't I use "variable that assures.."

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+It is also possible for the Jupyter notebook to set
+environmental variable that assures it will connect to an external Spark cluster.
+
+To do that you can optionally:
```
s/can optionally:/can optionally run this:/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+
+[[usage]]
+== Usage
+Both notebooks comes with some prepared data in the parquet files, but part of the
```
s/comes/come/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+Both notebooks comes with some prepared data in the parquet files, but part of the
+project is also the https://github.com/Jiri-Kremser/bitcoin-insights/tree/master/parquet-converter[parquet-converter]
+that takes the raw Blockchain data and converts
+it to the format that is digestible by Spark. It assumes the official bitcoin wallet to be installed
```
i'm not quite following the sentence starting with "It assumes", what did you mean to say here?

Jiri-Kremser replied:
it = the converter

the converter takes the files downloaded by the default Bitcoin client as its input and create parquet files with nodes and edges as the output. I wanted to mention that it's required for the converter that the official bitcoin client is installed (either `bin/bitcoin-qt` or `bin/bitcoind`). 

However, the docker images contain some pre-baked data so if user is not interested in some particular time period, it should just work. Otherwise, they should use the converter to prepare own data for the analysis. The makefile expects some env variables with the paths to the raw blk files (https://git.io/vFg01).

Would it be better if I put it this way:
"
The converter assumes the official bitcoin client to be installed, because it downloads the whole Blockchain (~150 GB) into `~/.bitcoin/blocks/`. Those files located in this directory are then the input to the converter, for more details see the [Makefile](https://git.io/vFg01).
"
?

elmiko replied:
yeah, i like the way you put it. i think it would also be worthwhile to explain a little more. 

this is good info imo
> However, the docker images contain some pre-baked data so if user is not interested in some particular time period, it should just work. Otherwise, they should use the converter to prepare own data for the analysis. The makefile expects some env variables with the paths to the raw blk files (https://git.io/vFg01).

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+that takes the raw Blockchain data and converts
+it to the format that is digestible by Spark. It assumes the official bitcoin wallet to be installed
+that itself downloads the whole Blockchain (~150 GB). However, it works also with
+partially downloaded Blockchain.
```
s/with partially/with a partially/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+in the toolbar or by pressing shift + enter. This notebook includes cells with
+instructions for running the program as well as the explanatory video about how the Blockchain internally works
++
+The description of what the notebook actually does should be part of the notebook itself.
```
s/does should be part/does is part/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
++
+The description of what the notebook actually does should be part of the notebook itself.
+In short this notebook demonstrates the degree distribution of the bitcoin addresses
+and compares with older Blockchain data. It also visualizes sub-graphs using sigma.js library and
```
s/using sigma.js/using the sigma.js/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+The description of what the notebook actually does should be part of the notebook itself.
+In short this notebook demonstrates the degree distribution of the bitcoin addresses
+and compares with older Blockchain data. It also visualizes sub-graphs using sigma.js library and
+with https://networkx.github.io/[NetworkX] python library.
```
s/with https/with the https/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+
+* Spark Notebook (GraphX and PageRank)
++
+While the previous notebook was trying to look at the graph as a whole, find
```
s/was trying to look at/was looking at/

s/find/finding/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+* Spark Notebook (GraphX and PageRank)
++
+While the previous notebook was trying to look at the graph as a whole, find
+some common properties and compare some statistics how they developed in time,
```
s/and compare/and comparing/

s/statistics how they developed in time/statistics about how the developed over time/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+While the previous notebook was trying to look at the graph as a whole, find
+some common properties and compare some statistics how they developed in time,
+this one demonstrates the famous PageRank algorithms and finds outliers in the graph and
+sort them by the significancy.
```
s/sort/sorts/

s/significancy/significance/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+[[expansion]]
+== Expansion
+Currently the number of edges in the graph is relatively large and it is due to
+the fact that during the conversion to parquet files we do following. For each
```
s/do following. For/do the following: for/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+Currently the number of edges in the graph is relatively large and it is due to
+the fact that during the conversion to parquet files we do following. For each
+transaction in each block we create an edge between each input in the transaction
+and each output in the transaction. This means that in case of multi-input (N),
```
s/that in case/that in the case/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+and each output in the transaction. This means that in case of multi-input (N),
+multi-output (M) transaction we end up with N times M edges.
+
+Useful improvement would be representing the concept of transaction in the resulting
```
s/Useful improvement/A useful improvement/

s/of transaction/of a transaction/

Jiri-Kremser commented:
@elmiko I hope I fixed them all, thanks a lot for being the grammar nazi :ok_hand:  I can't see those mistakes as a non-native speaker.

elmiko commented:
no worries Jirka, this a great effort and i don't mind helping get the language proper =)

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+a partially downloaded Blockchain.
+
+* Jupyter Notebook (GraphFrames and NetworkX)
++
```
i think you should add something here about the 2 pynb's included, i was a little confused when i saw "blockchain.ipynb" and "graphframes.ipynb". it would be nice to give the user a little advice, like "start with the blockchain nb then move on to graphframe" or similar.

Jiri-Kremser replied:
Mike, I removed the `graphframes.ipynb` from the image (https://github.com/Jiri-Kremser/bitcoin-insights/commit/efe5821a55736a2de00d3a562ca736126bd326dc) and add couple of sentences in https://github.com/radanalyticsio/radanalyticsio.github.io/pull/130/commits/72ad0b5455542d43d36eecbeae4d0de742b7bdb8

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -0,0 +1,107 @@
+with the https://networkx.github.io/[NetworkX] python library.
+
+* Spark Notebook (GraphX and PageRank)
++
```
same here, we should tell the user what notebook they will find inside

Jiri-Kremser commented:
> ImportError: No module named networkx

wow, not sure what happened, but this worked before. I've moved the module installation into the Dockerfile and now it works. The docker image is updated.

Also, I removed the `graphframes.ipynb` notebook completely, it is not related to the blockchain data at all and it's confusing.

The story behind the `graphframes.ipynb` was that I wanted to make the page rank algorithm working on graphframes, but it was failing even on very simple and small graphs, same for label propagation algorithm. The `graphframes.ipynb` served for me as a sand box notebook to experiment on very small data. The page rank works well in the graphX API, but fails with graphframes, but for pyspark there is only graphframes api, thats why we have two different notebooks: 1) using scala and graphx, for detecting the top 10 nodes by page rank and 2) using the python + graphframes for calculating global stastistics of the graph.

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -79,11 +81,16 @@ and compares with older Blockchain data. It also visualizes sub-graphs using the
 sorts them by the significance.
++
+It uses the Scala kernel so while the `blockchain.ipynb` used Python as the programming language,
+this one uses Scala. It is because the GraphX can be used only from Scala, while for GraphFrames there
```
s/It is/This is/

elmiko commented on _examples/blockchain.adoc:
```diff
@@ -79,11 +81,16 @@ and compares with older Blockchain data. It also visualizes sub-graphs using the
 this one demonstrates the famous PageRank algorithms and finds outliers in the graph and
 sorts them by the significance.
++
+It uses the Scala kernel so while the `blockchain.ipynb` used Python as the programming language,
```
s/kernel so/kernel, so/

elmiko commented:
i just tried running the jupyter notebook again and i'm still seeing the graphframes pynb, did the tag change on the image or something?

Jiri-Kremser commented:
@elmiko you are right, I pushed wrong image. Now, it should be ok.

willb commented:
@Jiri-Kremser no review comments but I wanted to get on the record as saying this is an awesome tutorial!  Thanks for doing the hard work to turn a great demo into a great addition to the website.

elmiko commented:
everything worked great Jirka, i'm going to merge this. thanks!
